pkg/nfd-topology-gc/nfd-nrt-gc.go

```diff
@@ -0,0 +1,219 @@
+	stopChan   chan struct{}
+	topoClient topologyclientset.Interface
+	gcPeriod   time.Duration
+	factory    informers.SharedInformerFactory
+}
+
+func New(args *Args) (NRTGarbageCollector, error) {
+
+	kubeconfig, err := apihelper.GetKubeconfig(args.Kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	stop := make(chan struct{})
+
+	return newNRTGarbageCollector(kubeconfig, stop, args.GCPeriod)
+}
+
+func newNRTGarbageCollector(config *restclient.Config, stop chan struct{}, gcPeriod time.Duration) (*nrtGarbageCollector, error) {
+	helper := apihelper.K8sHelpers{Kubeconfig: config}
+	cli, err := helper.GetTopologyClient()
+	if err != nil {
+		return nil, err
+	}
+
+	clientset := kubernetes.NewForConfigOrDie(config)
+	factory := informers.NewSharedInformerFactory(clientset, 5*time.Minute)
+
+	return &nrtGarbageCollector{
+		topoClient: cli,
+		stopChan:   stop,
+		gcPeriod:   gcPeriod,
+		factory:    factory,
+	}, nil
+}
+
+func (n *nrtGarbageCollector) deleteNodeHandler(object interface{}) {
+	// handle a case when we are starting up and need to clear stale NRT resources
+	obj := object
+	if deletedFinalStateUnknown, ok := object.(cache.DeletedFinalStateUnknown); ok {
+		klog.V(2).Infof("found stale NodeResourceTopology for node: %s ", object)
+		obj = deletedFinalStateUnknown.Obj
+	}
+
+	node, ok := obj.(*corev1.Node)
+	if !ok {
+		klog.Errorf("cannot convert %v to v1.Node", object)
+		return
+	}
+	if err := n.topoClient.TopologyV1alpha1().NodeResourceTopologies().Delete(context.TODO(), node.GetName(), metav1.DeleteOptions{}); err != nil {
+		if errors.IsNotFound(err) {
+			klog.V(2).Infof("NodeResourceTopology for node %s not found, omitting deletion", node.GetName())
+			return
+		} else {
+			klog.Warningf("failed to delete NodeResourceTopology for node %s: %s", node.GetName(), err.Error())
+			return
+		}
+	}
+	klog.Infof("NodeResourceTopology for node %s has been deleted", node.GetName())
+}
+
+// populateNodeIndexer populates cache with NRTs so we know which one to delete on first synchronization
+func (n *nrtGarbageCollector) populateNodeIndexer(indexer cache.Indexer) error {
+	nrts, err := n.topoClient.TopologyV1alpha1().NodeResourceTopologies().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, nrt := range nrts.Items {
+		node := &corev1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: nrt.GetName(),
+			},
+		}
+		klog.V(2).Infof("adding node %s to indexer", node.GetName())
+		err := indexer.Add(node)
+		if err != nil {
+			return err
+		}
+
+	}
+
+	return nil
+}
+
+// periodicGC runs garbage collector at every gcPeriod to make sure we haven't missed any node
+func (n *nrtGarbageCollector) periodicGC(gcPeriod time.Duration) {
+	indexer := n.factory.Core().V1().Nodes().Informer().GetIndexer()
+	gcTrigger := time.NewTicker(gcPeriod)
+	for {
+		select {
+		case <-gcTrigger.C:
+			klog.Infof("Running periodic GC")
+			objects := indexer.List()
+			nodes := map[string]struct{}{}
```

Comment:
(minor) use `k8s.io/apimachinery/pkg/util/sets` sets.String? we vendor` apimachinery/pkg/util anyway already

Reply:
Good idea!